File: Permutations/Permutations.cpp

class Solution {
public:
    void dfs(vector<vector<int> > &ans,vector<int> &num,int pos)
    {
        if(pos==num.size())
            ans.push_back(num);
        for(int i=pos;i<num.size();i++)
        {
            swap(num[pos],num[i]);
            dfs(ans,num,pos+1);
            swap(num[pos],num[i]);
        }
    }
    vector<vector<int> > permute(vector<int> &num) {
        vector<vector<int> >ans;
        dfs(ans,num,0);
        return ans;
    }
};


 

File: Permutations/Permutations.py


class Solution:
    def dfs(self,pos,num,ans,tempans):
        n=len(num)
        if pos==n:
            temp=tempans[:]
            ans.append(temp)
        for i in range(pos,n):
            num[pos],num[i]=num[i],num[pos]
            tempans.append(num[pos])
            self.dfs(pos+1,num,ans,tempans)
            tempans.pop()
            num[pos],num[i]=num[i],num[pos]
    # @param num, a list of integer
    # @return a list of lists of integers
    def permute(self, num):
        ans=[]
        tempans=[]
        self.dfs(0,num,ans,tempans)
        return ans
 